include/cell.tcc

```diff
@@ -949,3 +949,97 @@ Eigen::VectorXd mpm::Cell<Tdim>::interpolate_nodal_acceleration(
+
+      this->velocity_constraints_[face_id].emplace_back(
+          std::make_pair<unsigned, double>(static_cast<unsigned>(dir),
+                                           static_cast<double>(velocity)));
+    } else
+      throw std::runtime_error("Constraint direction is out of bounds");
+  } catch (std::exception& exception) {
+    console_->error("{} #{}: {}\n", __FILE__, __LINE__, exception.what());
+    status = false;
+  }
+  return status;
+}
+
+//! Apply velocity constraints
+template <unsigned Tdim>
+void mpm::Cell<Tdim>::apply_velocity_constraints() {
+
+  // TODO
+}
+
+//! Compute all face normal 2d
+template <>
+inline void mpm::Cell<2>::compute_normals() {
+
+  //! Set number of faces from element
+  for (unsigned face_id = 0; face_id < element_->nfaces(); ++face_id) {
+    // Get the nodes of the face
+    Eigen::VectorXi indices = element_->face_indices(face_id);
+
+    // Compute the vector to calculate normal (perpendicular)
+    // a = node(0) - node(1)
+    Eigen::Matrix<double, 2, 1> a = (this->nodes_[indices(0)])->coordinates() -
+                                    (this->nodes_[indices(1)])->coordinates();
+
+    // Compute normal and make unit vector
+    // The normal vector n to vector a is defined such that the dot product
+    // between a and n is always 0 In 2D, n(0) = -a(1), n(1) = a(0) Note that
+    // the reverse does not work to produce normal that is positive pointing out
+    // of the element
+    Eigen::Matrix<double, 2, 1> normal_vector;
+    normal_vector(0) = -a(1);
+    normal_vector(1) = a(0);
+    normal_vector = normal_vector.normalized();
+
+    // Store to private variable
+    face_normals_.insert(std::make_pair<unsigned, Eigen::VectorXd>(
+        static_cast<unsigned>(face_id), normal_vector));
+  }
+}
+
+//! Compute all face normal 3d
+template <>
+inline void mpm::Cell<3>::compute_normals() {
+
+  //! Set number of faces from element
+  for (unsigned face_id = 0; face_id < element_->nfaces(); ++face_id) {
+    // Get the nodes of the face
+    Eigen::VectorXi indices = element_->face_indices(face_id);
+
+    // Compute two vectors to calculate normal
+    // a = node(1) - node(0)
+    // b = node(3) - node(0)
+    Eigen::Matrix<double, 3, 1> a = (this->nodes_[indices(1)])->coordinates() -
+                                    (this->nodes_[indices(0)])->coordinates();
+    Eigen::Matrix<double, 3, 1> b = (this->nodes_[indices(3)])->coordinates() -
+                                    (this->nodes_[indices(0)])->coordinates();
+
+    // Compute normal and make unit vector
+    // normal = a x b
+    // Note that definition of a and b are such that normal is always out of
+    // page
+    Eigen::Matrix<double, 3, 1> normal_vector = a.cross(b);
+    normal_vector = normal_vector.normalized();
+
+    // Store to private variable
+    face_normals_.insert(std::make_pair<unsigned, Eigen::VectorXd>(
+        static_cast<unsigned>(face_id), normal_vector));
```

Comment:
why do you want to static_cast something which is already an unsigned?

Reply:
for some reason it does not work when I don't do `static_cast` :(